lib/Parse/ParseType.cpp

```diff
@@ -1612,71 +1605,65 @@ bool Parser::canParseSimpleTypeIdentifier() {
   consumeToken();
 
   // Parse an optional generic argument list.
-  if (startsWithLess(Tok))
-    if (!canParseGenericArguments())
-      return false;
+  if (startsWithLess(Tok) && !canParseGenericArguments())
+    return false;
 
   return true;
 }
 
 bool Parser::canParseTypeIdentifier() {
-  while (true) {
+  for (;;) {
     if (!canParseSimpleTypeIdentifier())
       return false;
 
     // Treat 'Foo.<anything>' as an attempt to write a dotted type
     // unless <anything> is 'Type' or 'Protocol'.
-    if ((Tok.is(tok::period) || Tok.is(tok::period_prefix)) &&
-        !peekToken().isContextualKeyword("Type") &&
-        !peekToken().isContextualKeyword("Protocol")) {
-      consumeToken();
-    } else {
+    if (!(Tok.is(tok::period) || Tok.is(tok::period_prefix)) ||
+        peekToken().isContextualKeyword("Type") ||
+        peekToken().isContextualKeyword("Protocol")) {
       return true;
     }
+    consumeToken();
   }
 }
 
 bool Parser::canParseBaseTypeForQualifiedDeclName() {
   BacktrackingScope backtrack(*this);
 
   // Parse a simple type identifier.
-  if (!canParseSimpleTypeIdentifier())
-    return false;
-
   // Qualified name base types must be followed by a period.
   // If the next token starts with a period, return true.
-  return startsWithSymbol(Tok, '.');
+  return canParseSimpleTypeIdentifier() && startsWithSymbol(Tok, '.');
```

Comment:
Here too, the comments no longer describe the code. Please restore the original code or update the comments.